### PR TITLE
sync: add merge() to semaphore permits

### PR DESCRIPTION
**Pull request by domodwyer**
An implementation of Semaphore permit merging, as described in #4947.

---

* sync: add merge() to semaphore permits (d8afc6ac)

      Adds a merge() method to:

         * SemaphorePermit
         * OwnedSemaphorePermit

      Merging two permits instances together consumes one instance, adding the 
      permits it holds to the remaining instance.

### CODE REVIEW

File: tokio/src/sync/semaphore.rs [hunk 1]
```diff
@@ -620,6 +620,19 @@ impl<'a> SemaphorePermit<'a> {
     pub fn forget(mut self) {
         self.permits = 0;
     }
+
+    /// Merge two [`SemaphorePermit`] instances together, consuming `other`
+    /// without releasing the permits it holds.
+    ///
+    /// Permits held by both `self` and `other` are released when `self` drops.
```
Comment (cynecx) [inline tokio/src/sync/semaphore.rs:627]: The docs should mention the conditions when `merge` will panic.

Comment (Darksonn) [reply tokio/src/sync/semaphore.rs:627]: Additionally, the method should be marked `#[track_caller]`.

Comment (domodwyer) [reply tokio/src/sync/semaphore.rs:627]: Done in 5900f845a418f52c12837fec8acd29923ce827bb - thanks! :+1:

**Review by Darksonn — COMMENTED** [review]
Can you add a test to `tokio/tests/sync_panic.rs` too?

**Comment by domodwyer** [comment]
I'm not really sure I understand what's happening with the failing wasm test: https://github.com/tokio-rs/tokio/runs/8118383257?check_suite_focus=true

Does `#[should_panic]` not catch panics on wasm targets?

**Comment by Darksonn** [comment]
Unwinding doesn't work on wasm.

**Comment by domodwyer** [comment]
> Unwinding doesn't work on wasm.

Yes, what is the profered way to proceed? `#[cfg(...)]` out the test for the WASM platform?

**Comment by Darksonn** [comment]
Yes.

**Comment by domodwyer** [comment]
Everything is green, thanks for your help @Darksonn :+1:

**Review by Darksonn — APPROVED** [review]
LGTM.

**Pull request merged**